test/python/algorithms/test_vqd.py

```diff
@@ -0,0 +1,584 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2018, 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+""" Test VQD """
+
+import logging
+import unittest
+from test.python.algorithms import QiskitAlgorithmsTestCase
+
+import numpy as np
+from ddt import data, ddt, unpack
+
+from qiskit import BasicAer, QuantumCircuit
+from qiskit.algorithms import VQD, AlgorithmError
+from qiskit.algorithms.optimizers import (
+    COBYLA,
+    L_BFGS_B,
+    SLSQP,
+)
+from qiskit.circuit.library import EfficientSU2, RealAmplitudes, TwoLocal
+from qiskit.exceptions import MissingOptionalLibraryError
+from qiskit.opflow import (
+    AerPauliExpectation,
+    I,
+    MatrixExpectation,
+    MatrixOp,
+    PauliExpectation,
+    PauliSumOp,
+    PrimitiveOp,
+    X,
+    Z,
+)
+
+from qiskit.utils import QuantumInstance, algorithm_globals, has_aer
+from ..transpiler._dummy_passes import DummyAP
+
+if has_aer():
+    from qiskit import Aer
+
+logger = "LocalLogger"
+
+
+class LogPass(DummyAP):
+    """A dummy analysis pass that logs when executed"""
+
+    def __init__(self, message):
+        super().__init__()
+        self.message = message
+
+    def run(self, dag):
+        logging.getLogger(logger).info(self.message)
+
+
+@ddt
+class TestVQD(QiskitAlgorithmsTestCase):
```

Comment:
Should we add some tests for strange values of `k` a user might support? Like 0, -10, or 10000?